For various reasons sometimes the debcargo-conf patches can't be applied as-is,
so we further patch them here. The reasons are as follows:

- commoncrypto, commoncrypto-sys: these crates were released before crates.io
  automatically re-wrote all Cargo.toml files. cargo-vendor outputs the
  un-re-written original Cargo.toml files, but debcargo will forcibly rewrite
  even un-re-written ones that were published to crates.io. The patch below
  rewrites our existing patches so they work against the un-re-written ones.

- unicode-normalization, vte, tempfile: have dependency relaxation that applies,
  but doesn't match the vendored copy

- concolor: more recent version in debcargo-conf, can be dropped once cargo
  upstream updates

- clap: clap is v4 in debcargo-conf, patches from clap-3 temporarily patched in
  until we update to a cargo version using clap 4.x

diff --git a/src/clap/debian/patches/disable-derive-deprecated.patch b/src/clap/debian/patches/disable-derive-deprecated.patch
new file mode 100644
index 000000000..6ed82171f
--- /dev/null
+++ b/src/clap/debian/patches/disable-derive-deprecated.patch
@@ -0,0 +1,11 @@
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -418,7 +418,7 @@
+     "color",
+     "suggestions",
+ ]
+-deprecated = ["clap_derive/deprecated"]
++deprecated = []
+ derive = [
+     "clap_derive",
+     "once_cell",
diff --git a/src/clap/debian/patches/disable-snapbox.patch b/src/clap/debian/patches/disable-snapbox.patch
index a6fbd083a..9335518c4 100644
--- a/src/clap/debian/patches/disable-snapbox.patch
+++ b/src/clap/debian/patches/disable-snapbox.patch
@@ -1,6 +1,12 @@
 --- a/Cargo.toml
 +++ b/Cargo.toml
-@@ -400,3 +399,0 @@
+@@ -397,9 +397,6 @@
+ [dev-dependencies.shlex]
+ version = "1.1.0"
+ 
 -[dev-dependencies.snapbox]
--version = "0.4"
+-version = "0.2.9"
 -
+ [dev-dependencies.static_assertions]
+ version = "1.1.0"
+ 
diff --git a/src/clap/debian/patches/disable-trycmd.diff b/src/clap/debian/patches/disable-trycmd.diff
index aeb0bbad4..053bf14f9 100644
--- a/src/clap/debian/patches/disable-trycmd.diff
+++ b/src/clap/debian/patches/disable-trycmd.diff
@@ -1,8 +1,11 @@
 --- a/Cargo.toml
 +++ b/Cargo.toml
-@@ -413,9 +412,0 @@
+@@ -410,15 +410,6 @@
+ [dev-dependencies.trybuild]
+ version = "1.0.18"
+ 
 -[dev-dependencies.trycmd]
--version = "0.14.3"
+-version = "0.13"
 -features = [
 -    "color-auto",
 -    "diff",
@@ -10,3 +13,6 @@
 -]
 -default-features = false
 -
+ [features]
+ cargo = ["once_cell"]
+ color = [
diff --git a/src/clap/debian/patches/disable-unic-emoji-char.patch b/src/clap/debian/patches/disable-unic-emoji-char.patch
deleted file mode 100644
index 97aa25af0..000000000
--- a/src/clap/debian/patches/disable-unic-emoji-char.patch
+++ /dev/null
@@ -1,36 +0,0 @@
-Index: clap/Cargo.toml
-===================================================================
---- clap.orig/Cargo.toml
-+++ clap/Cargo.toml
-@@ -408,9 +408,6 @@ version = "1.1.0"
- [dev-dependencies.trybuild]
- version = "1.0.71"
- 
--[dev-dependencies.unic-emoji-char]
--version = "0.9.0"
--
- [features]
- cargo = ["dep:once_cell"]
- color = [
-Index: clap/src/output/textwrap/core.rs
-===================================================================
---- clap.orig/src/output/textwrap/core.rs
-+++ clap/src/output/textwrap/core.rs
-@@ -90,7 +90,7 @@ mod tests {
-     #[cfg(feature = "unicode")]
-     use unicode_width::UnicodeWidthChar;
- 
--    #[test]
-+    /*#[test]
-     fn emojis_have_correct_width() {
-         use unic_emoji_char::is_emoji;
- 
-@@ -129,7 +129,7 @@ mod tests {
- 
-         // The remaining planes contain almost no assigned code points
-         // and thus also no emojis.
--    }
-+    }*/
- 
-     #[test]
-     #[cfg(feature = "unicode")]
diff --git a/src/clap/debian/patches/disable-unstable-doc.patch b/src/clap/debian/patches/disable-unstable-doc.patch
index 00de9ce79..404f1d9cf 100644
--- a/src/clap/debian/patches/disable-unstable-doc.patch
+++ b/src/clap/debian/patches/disable-unstable-doc.patch
@@ -1,25 +1,38 @@
 --- a/Cargo.toml
 +++ b/Cargo.toml
-@@ -39,3 +39,2 @@
+@@ -37,7 +37,6 @@
+ repository = "https://github.com/clap-rs/clap"
+ 
  [package.metadata.docs.rs]
 -features = ["unstable-doc"]
  rustdoc-args = [
-@@ -49,5 +48,2 @@
+     "--cfg",
+     "docsrs",
+@@ -47,9 +46,6 @@
+     "-Zrustdoc-scrape-examples=examples",
+ ]
  
 -[package.metadata.playground]
 -features = ["unstable-doc"]
 -
  [package.metadata.release]
-@@ -439,12 +435,2 @@
+ shared-version = true
+ tag-name = "v{{version}}"
+@@ -437,17 +433,6 @@
+     "textwrap/unicode-width",
+     "unicase",
  ]
 -unstable-doc = [
 -    "derive",
 -    "cargo",
 -    "wrap_help",
+-    "yaml",
 -    "env",
 -    "unicode",
--    "string",
+-    "regex",
 -    "unstable-replace",
 -    "unstable-grouped",
 -]
  unstable-grouped = []
+ unstable-replace = []
+ unstable-v4 = [
diff --git a/src/clap/debian/patches/disable-unstable-v4.patch b/src/clap/debian/patches/disable-unstable-v4.patch
new file mode 100644
index 000000000..2993a959a
--- /dev/null
+++ b/src/clap/debian/patches/disable-unstable-v4.patch
@@ -0,0 +1,13 @@
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -435,10 +435,6 @@
+ ]
+ unstable-grouped = []
+ unstable-replace = []
+-unstable-v4 = [
+-    "clap_derive/unstable-v4",
+-    "deprecated",
+-]
+ wrap_help = [
+     "terminal_size",
+     "textwrap/terminal_size",
diff --git a/src/clap/debian/patches/disable-unstable-v5.patch b/src/clap/debian/patches/disable-unstable-v5.patch
deleted file mode 100644
index 5c2a59a30..000000000
--- a/src/clap/debian/patches/disable-unstable-v5.patch
+++ /dev/null
@@ -1,7 +0,0 @@
---- a/Cargo.toml
-+++ b/Cargo.toml
-@@ -438,4 +437,0 @@
--unstable-v5 = [
--    "clap_derive?/unstable-v5",
--    "deprecated",
--]
diff --git a/src/clap/debian/patches/once-cell-non-optional.patch b/src/clap/debian/patches/once-cell-non-optional.patch
new file mode 100644
index 000000000..0e6692e81
--- /dev/null
+++ b/src/clap/debian/patches/once-cell-non-optional.patch
@@ -0,0 +1,29 @@
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -354,7 +354,6 @@
+ 
+ [dependencies.once_cell]
+ version = "1.12.0"
+-optional = true
+ 
+ [dependencies.regex]
+ version = "1.0"
+@@ -404,7 +403,7 @@
+ version = "1.0.18"
+ 
+ [features]
+-cargo = ["once_cell"]
++cargo = []
+ color = [
+     "atty",
+     "termcolor",
+@@ -420,8 +419,7 @@
+ ]
+ deprecated = []
+ derive = [
+-    "clap_derive",
+-    "once_cell",
++    "clap_derive"
+ ]
+ env = []
+ std = ["indexmap/std"]
diff --git a/src/clap/debian/patches/revert-switch-to-is-terminal.patch b/src/clap/debian/patches/revert-switch-to-is-terminal.patch
deleted file mode 100644
index 425ff3aca..000000000
--- a/src/clap/debian/patches/revert-switch-to-is-terminal.patch
+++ /dev/null
@@ -1,44 +0,0 @@
-Index: clap/src/output/fmt.rs
-===================================================================
---- clap.orig/src/output/fmt.rs
-+++ clap/src/output/fmt.rs
-@@ -82,9 +82,10 @@ impl std::fmt::Display for Colorizer {
- 
- #[cfg(feature = "color")]
- fn is_a_tty(stream: Stream) -> bool {
--    use is_terminal::IsTerminal;
--    match stream {
--        Stream::Stdout => std::io::stdout().is_terminal(),
--        Stream::Stderr => std::io::stderr().is_terminal(),
--    }
-+    let stream = match stream {
-+        Stream::Stdout => atty::Stream::Stdout,
-+        Stream::Stderr => atty::Stream::Stderr,
-+    };
-+
-+    atty::is(stream)
- }
-Index: clap/Cargo.toml
-===================================================================
---- clap.orig/Cargo.toml
-+++ clap/Cargo.toml
-@@ -365,8 +365,8 @@ optional = true
- [dependencies.clap_lex]
- version = "0.3.0"
- 
--[dependencies.is-terminal]
--version = "0.4.1"
-+[dependencies.atty]
-+version = "0.2"
- optional = true
- 
- [dependencies.once_cell]
-@@ -414,7 +414,7 @@ version = "0.9.0"
- [features]
- cargo = ["dep:once_cell"]
- color = [
--    "dep:is-terminal",
-+    "dep:atty",
-     "dep:termcolor",
- ]
- debug = [
diff --git a/src/clap/debian/patches/series b/src/clap/debian/patches/series
index 137658685..47bdb2c6b 100644
--- a/src/clap/debian/patches/series
+++ b/src/clap/debian/patches/series
@@ -1,6 +1,6 @@
 disable-trycmd.diff
 disable-unstable-doc.patch
-disable-unstable-v5.patch
+disable-unstable-v4.patch
 disable-snapbox.patch
-revert-switch-to-is-terminal.patch
-disable-unic-emoji-char.patch
+disable-derive-deprecated.patch
+once-cell-non-optional.patch
diff --git a/src/commoncrypto-sys/debian/patches/no-clippy.patch b/src/commoncrypto-sys/debian/patches/no-clippy.patch
index 3e4c8850e..93789f3b8 100644
--- a/src/commoncrypto-sys/debian/patches/no-clippy.patch
+++ b/src/commoncrypto-sys/debian/patches/no-clippy.patch
@@ -1,17 +1,16 @@
---- a/Cargo.toml	2017-01-24 06:56:51.000000000 +0000
-+++ b/Cargo.toml	2018-09-15 15:37:47.602333479 +0000
-@@ -19,14 +19,8 @@
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -8,13 +8,8 @@
  keywords = ["crypto", "hash", "digest", "osx", "commoncrypto"]
  license = "MIT"
- repository = "https://github.com/malept/rust-commoncrypto"
--[dependencies.clippy]
--version = "0.0"
--optional = true
  
- [dependencies.libc]
- version = "0.2"
- [dev-dependencies.hex]
- version = "0.2"
--
 -[features]
 -lint = ["clippy"]
+-
+ [dependencies]
+ libc = "0.2"
+ 
+-clippy = { version = "0.0", optional = true }
+-
+ [dev-dependencies]
+ hex = "0.2"
diff --git a/src/commoncrypto/debian/patches/no-clippy.patch b/src/commoncrypto/debian/patches/no-clippy.patch
index 38d9c9255..b21a7cae5 100644
--- a/src/commoncrypto/debian/patches/no-clippy.patch
+++ b/src/commoncrypto/debian/patches/no-clippy.patch
@@ -1,17 +1,16 @@
---- a/Cargo.toml	2017-01-24 06:56:51.000000000 +0000
-+++ b/Cargo.toml	2018-09-15 17:37:01.354423224 +0000
-@@ -19,14 +19,8 @@
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -8,13 +8,8 @@
  keywords = ["crypto", "hash", "digest", "osx", "commoncrypto"]
  license = "MIT"
- repository = "https://github.com/malept/rust-commoncrypto"
--[dependencies.clippy]
--version = "0.0"
--optional = true
  
- [dependencies.commoncrypto-sys]
- version = "0.2.0"
- [dev-dependencies.hex]
- version = "0.2"
--
 -[features]
 -lint = ["clippy"]
+-
+ [dependencies]
+ commoncrypto-sys = { version = "0.2.0", path = "../commoncrypto-sys" }
+ 
+-clippy = { version = "0.0", optional = true }
+-
+ [dev-dependencies]
+ hex = "0.2"
diff --git a/src/concolor/debian/patches/series b/src/concolor/debian/patches/series
index cdf075927..942673b44 100644
--- a/src/concolor/debian/patches/series
+++ b/src/concolor/debian/patches/series
@@ -1,2 +1,2 @@
 avoid-doctest-breakage.patch
-switch-back-to-atty.patch
+#switch-back-to-atty.patch
diff --git a/src/tempfile/debian/patches/series b/src/tempfile/debian/patches/series
index 75abb7f14..2378a84ba 100644
--- a/src/tempfile/debian/patches/series
+++ b/src/tempfile/debian/patches/series
@@ -1 +1 @@
-relax-dep.diff
+#relax-dep.diff
diff --git a/src/unicode-normalization/debian/patches/series b/src/unicode-normalization/debian/patches/series
index 2c1db10bd..39fe7a4da 100644
--- a/src/unicode-normalization/debian/patches/series
+++ b/src/unicode-normalization/debian/patches/series
@@ -1,2 +1 @@
 disable-tests-missing-testdata.patch
-switch-back-to-smallvec.patch
diff --git a/src/unicode-normalization/debian/patches/switch-back-to-smallvec.patch b/src/unicode-normalization/debian/patches/switch-back-to-smallvec.patch
deleted file mode 100644
index 06e816ca7..000000000
--- a/src/unicode-normalization/debian/patches/switch-back-to-smallvec.patch
+++ /dev/null
@@ -1,134 +0,0 @@
-This patch switches back to using smallvec as we do not have tinyvec in Debian.
-It is based on a revert of upstream commit d6abe8e5148c2c513cb65f487a037c7600de2498
-adapted by Peter Michael Green for use in the Debian package.
-
-Index: unicode-normalization/src/decompose.rs
-===================================================================
---- unicode-normalization.orig/src/decompose.rs
-+++ unicode-normalization/src/decompose.rs
-@@ -10,7 +10,7 @@
- use core::fmt::{self, Write};
- use core::iter::Fuse;
- use core::ops::Range;
--use tinyvec::TinyVec;
-+use smallvec::SmallVec;
- 
- #[derive(Clone)]
- enum DecompositionType {
-@@ -32,7 +32,7 @@ pub struct Decompositions<I> {
-     // 2) "Ready" characters which are sorted and ready to emit on demand;
-     // 3) A "pending" block which stills needs more characters for us to be able
-     //    to sort in canonical order and is not safe to emit.
--    buffer: TinyVec<[(u8, char); 4]>,
-+    buffer: SmallVec<[(u8, char); 4]>,
-     ready: Range<usize>,
- }
- 
-@@ -41,7 +41,7 @@ pub fn new_canonical<I: Iterator<Item =
-     Decompositions {
-         kind: self::DecompositionType::Canonical,
-         iter: iter.fuse(),
--        buffer: TinyVec::new(),
-+        buffer: SmallVec::new(),
-         ready: 0..0,
-     }
- }
-@@ -51,7 +51,7 @@ pub fn new_compatible<I: Iterator<Item =
-     Decompositions {
-         kind: self::DecompositionType::Compatible,
-         iter: iter.fuse(),
--        buffer: TinyVec::new(),
-+        buffer: SmallVec::new(),
-         ready: 0..0,
-     }
- }
-Index: unicode-normalization/src/lib.rs
-===================================================================
---- unicode-normalization.orig/src/lib.rs
-+++ unicode-normalization/src/lib.rs
-@@ -50,7 +50,7 @@ extern crate alloc;
- #[cfg(feature = "std")]
- extern crate core;
- 
--extern crate tinyvec;
-+extern crate smallvec;
- 
- pub use crate::decompose::Decompositions;
- pub use crate::quick_check::{
-Index: unicode-normalization/src/recompose.rs
-===================================================================
---- unicode-normalization.orig/src/recompose.rs
-+++ unicode-normalization/src/recompose.rs
-@@ -10,7 +10,7 @@
- 
- use crate::decompose::Decompositions;
- use core::fmt::{self, Write};
--use tinyvec::TinyVec;
-+use smallvec::SmallVec;
- 
- #[derive(Clone)]
- enum RecompositionState {
-@@ -24,7 +24,7 @@ enum RecompositionState {
- pub struct Recompositions<I> {
-     iter: Decompositions<I>,
-     state: RecompositionState,
--    buffer: TinyVec<[char; 4]>,
-+    buffer: SmallVec<[char; 4]>,
-     composee: Option<char>,
-     last_ccc: Option<u8>,
- }
-@@ -34,7 +34,7 @@ pub fn new_canonical<I: Iterator<Item =
-     Recompositions {
-         iter: super::decompose::new_canonical(iter),
-         state: self::RecompositionState::Composing,
--        buffer: TinyVec::new(),
-+        buffer: SmallVec::new(),
-         composee: None,
-         last_ccc: None,
-     }
-@@ -45,7 +45,7 @@ pub fn new_compatible<I: Iterator<Item =
-     Recompositions {
-         iter: super::decompose::new_compatible(iter),
-         state: self::RecompositionState::Composing,
--        buffer: TinyVec::new(),
-+        buffer: SmallVec::new(),
-         composee: None,
-         last_ccc: None,
-     }
-Index: unicode-normalization/Cargo.toml
-===================================================================
---- unicode-normalization.orig/Cargo.toml
-+++ unicode-normalization/Cargo.toml
-@@ -43,9 +43,8 @@ keywords = [
- license = "MIT/Apache-2.0"
- repository = "https://github.com/unicode-rs/unicode-normalization"
- 
--[dependencies.tinyvec]
-+[dependencies.smallvec]
- version = "1"
--features = ["alloc"]
- 
- [features]
- default = ["std"]
-Index: unicode-normalization/src/replace.rs
-===================================================================
---- unicode-normalization.orig/src/replace.rs
-+++ unicode-normalization/src/replace.rs
-@@ -8,7 +8,7 @@
- // option. This file may not be copied, modified, or distributed
- // except according to those terms.
- use core::fmt::{self, Write};
--use tinyvec::ArrayVec;
-+use smallvec::SmallVec;
- 
- /// External iterator for replacements for a string's characters.
- #[derive(Clone)]
-@@ -36,7 +36,7 @@ impl<I: Iterator<Item = char>> Iterator
-         match self.iter.next() {
-             Some(ch) => {
-                 // At this time, the longest replacement sequence has length 2.
--                let mut buffer = ArrayVec::<[char; 2]>::new();
-+                let mut buffer = SmallVec::<[char; 2]>::new();
-                 super::char::decompose_cjk_compat_variants(ch, |d| buffer.push(d));
-                 self.buffer = buffer.get(1).copied();
-                 Some(buffer[0])
diff --git a/src/vte/debian/patches/arrayvec.diff b/src/vte/debian/patches/arrayvec.diff
deleted file mode 100644
index 1923f9577..000000000
--- a/src/vte/debian/patches/arrayvec.diff
+++ /dev/null
@@ -1,26 +0,0 @@
-Index: vte/Cargo.toml
-===================================================================
---- vte.orig/Cargo.toml
-+++ vte/Cargo.toml
-@@ -24,7 +24,7 @@ categories = ["parsing", "no-std"]
- license = "Apache-2.0 OR MIT"
- repository = "https://github.com/alacritty/vte"
- [dependencies.arrayvec]
--version = "0.5.1"
-+version = "0.7"
- optional = true
- default-features = false
- 
-Index: vte/src/lib.rs
-===================================================================
---- vte.orig/src/lib.rs
-+++ vte/src/lib.rs
-@@ -78,7 +78,7 @@ pub struct Parser {
-     params: Params,
-     param: u16,
-     #[cfg(feature = "no_std")]
--    osc_raw: ArrayVec<[u8; MAX_OSC_RAW]>,
-+    osc_raw: ArrayVec<u8, MAX_OSC_RAW>,
-     #[cfg(not(feature = "no_std"))]
-     osc_raw: Vec<u8>,
-     osc_params: [(usize, usize); MAX_OSC_PARAMS],
diff --git a/src/vte/debian/patches/series b/src/vte/debian/patches/series
index 011b1a629..203bdd762 100644
--- a/src/vte/debian/patches/series
+++ b/src/vte/debian/patches/series
@@ -1,2 +1 @@
 00-remove-nightly-feature
-arrayvec.diff
